avail/src/main/kotlin/avail/interpreter/PojoClassLoader.kt

```diff
@@ -0,0 +1,207 @@
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package avail.interpreter
+
+import avail.builder.ModuleName
+import avail.descriptor.tuples.A_String
+import java.io.File
+import java.net.URLClassLoader
+import java.util.jar.JarFile
+import javax.annotation.concurrent.GuardedBy
+
+/**
+ * The [URLClassLoader] for loading external Jars through the Pojo linking
+ * system.
+ *
+ * @author Richard Arriaga
+ *
+ * @property moduleName
+ *   The fully qualified [A_String] name of the module that is responsible for
+ *   creating this [PojoClassLoader] for loading the target jar file.
+ *
+ * @constructor
+ * Construct a [PojoClassLoader].
+ *
+ * @param jarFile
+ *   The [File] location of the jar file to load.
+ * @param moduleName
+ *   The [A_String] module name of the module that created this class loader.
+ * @param parent
+ *   The parent [ClassLoader] of this [PojoClassLoader].
+ */
+class PojoClassLoader constructor(
+	jarFile: File,
+	val moduleName: A_String,
+	parent: ClassLoader = Primitive::class.java.classLoader
+): URLClassLoader(arrayOf(jarFile.toURI().toURL()), parent)
+{
+	/**
+	 * The set of [Primitive.PrimitiveHolder]s that were loaded by this
+	 * [PrimitiveClassLoader].
+	 */
+	private val holders = mutableSetOf<PojoHolder>()
+
+	/**
+	 * The path to the linked Jar file.
+	 */
+	private val jarPath = jarFile.path
+
+	init
+	{
+		try
+		{
+			JarFile(jarFile).use { jar ->
+				jar.entries().asIterator().forEach { entry ->
+					if (!entry.name.endsWith(".class")) return@forEach
+					entry.name.replace(".class", "").let {
+						val c = it.replace("/", ".")
+						val holder = PojoHolder(c, this)
+						PojoHolder.holdersByClassName[c] = holder
+						holders.add(holder)
+					}
+				}
+			}
+		}
+		catch (e: Throwable)
+		{
+			// We don't care what the exception is here, we just need to
+			// rollback the classes added to PojoHolder.holdersByClassName. The
+			// caller is responsible for handling said exceptions.
+			holders.forEach {
+				PojoHolder.holdersByClassName.remove(it.className)
+				throw e
+			}
+		}
+		moduleToLoader.computeIfAbsent(moduleName) { mutableSetOf() }.add(this)
+		jarToModule[jarPath] = moduleName
+	}
+
+	/**
+	 * Cleanup all of the [PojoHolder(it, this)]s loaded by this
```

Comment:
Pretty sure that this is a search-and-replace error. I think it's just supposed to be `[PojoHolder]s`, unless you know about some fancy feature of KDoc that has thus far eluded me.